File: frontend/src/game/upgrades/stat_modifiers/tenacity_3.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const Tenacity3Def: UpgradeDef = {
  id: "tenacity_3",
  name: "Tenacity",
  description: "Increased the bullet lifespan by 1 second",
  rarity: RarityID.Legendary,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 40,
  targetClass: UpgradeTargetID.Bullet,
  fieldInTargetClass: UpgradeStatID.TimeLeft,
  value: 1000,
  specificAttackType: "bullet",
  stackable: true,
  maxStacks: 1,
}

export class Tenacity3 extends Upgrade {}

File: backend/app/core/upgrades/stat_modifiers/tenacity_3.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class Tenacity3(UpgradeImplementation):
    id = 'tenacity_3'
    name = 'Tenacity'
    description = 'Increased the bullet lifespan by 1 second'
    rarity = 'legendary'
    type = 'stat_modifier'
    cost = 40

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 1000
        appliers.modifiers.add_multiplier('bullet', 'timeLeft', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True
